---
title: "TCMR <br>A Package for TCM research"
author:
  name: "<b>Saisai Tian</b>"
  affiliation: "<b>School of Pharmacy, <br>Second Military Medical University, <br>Shanghai, China</b>"
  email: "<b>372479584@qq.com</b>"
date: "<b>`r Sys.Date()`</b>"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
    toc: yes
vignette: >
  %\VignetteIndexEntry{TCMR-VIGNETTE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
<style>
p.caption {
  font-size: 0.8em;
}

sup {
    line-height: 0;
    font-size: 0.83em;
    vertical-align: super;
}

.math {
  font-size: small;
}
</style>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, echo=FALSE}
Sys.setenv(LANGUAGE = "en")
```

# <a id="Section.1" style="color:#159957;">1. Introduction</a>

Recent advances in next-generation sequencing, microarrays and mass spectrometry for omics data production have enabled the generation and collection of different modalities of high-dimensional molecular data<sup>1</sup>. Clustering multi-omic data has the potential to reveal further systems-level insights, but raises computational and biological challenges<sup>2</sup>. This vignette aims to show how to use the package named MOVICS to perform multi-omics integrative clustering and visualization for cancer subtyping researches. This R package provides a unified interface for 10 state-of-the-art multi-omics clustering algorithms, and standardizes the output for each algorithm so as to form a pipeline for downstream analyses. Ten algorithms are CIMLR, iClusterBayes, MoCluster, COCA, ConsensusClustering, IntNMF, LRAcluster, NEMO, PINSPlus, and SNF where the former three methods can also perform the process of feature selection. For cancer subtyping studies, MOVICS also forms a pipeline for most commonly used downstream analyses for further subtype characterization and creates editable publication-quality illustrations (see more details below). Please note that MOVICS currently supports up to 6 omics data for jointly clustering and users must provide at least 2 omics datasets as input. Okay then, let us make our hands dirty to figure out how MOVICS works.

# <a id="Section.2" style="color:#159957;">2. Installation</a>
It is essential that you have R 4.0.1 or above already installed on your computer or server. MOVICS is a pipeline that utilizes many other R packages that are currently available from CRAN, Bioconductor and GitHub. For all of the steps of the pipeline to work, make sure that you have upgraded Bioconductor to newest version (BiocManager v3.11).
After you have R and Bioconductor installed properly, you can start to install MOVICS. The easiest way to install it is by typing the following code into your R session:
```{r, warning=FALSE, message=FALSE, echo=TRUE, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!require("devtools")) 
    install.packages("devtools")
#devtools::install_github("saisaitian/TCMR")
```

When you are installing MOVICS, you may encounter some errors saying that some packages are not installed. These errors are caused by recursively depending on R packages, so if one package was not installed properly on your computer, MOVICS would fail. To solve these errors, you simply need to check those error messages, find out which packages required are missing, then install it with command `BiocManager::install("YourErrorPackage")` or `install.packages("YourErrorPackage")` or `devtools::install_github("username/YourErrorPackage")` directly. After that, retry installing MOVICS, it may take several times, but eventually it should work. Or, we highly suggest that you referred to the `Imports` in the [DESCRIPTION](https://github.com/xlucpu/MOVICS/blob/master/DESCRIPTION) file, try to install all the R dependencies, and then install MOVICS. We also summarized several common problems that you may meet when installing MOVICS, please refer to [Troubleshooting](https://github.com/xlucpu/MOVICS) if applicable. After installation, you should be able to load the MOVICS package in your R session:
```{r, eval=TRUE, warning=FALSE, message=FALSE}
library("TCMR")
library("dplyr")
library("knitr")
library(kableExtra)
```

# <a id="Section.3" style="color:#159957;">3. Real Data Scenario</a>
This package contains two pre-processed real expression datasets of TCMs on breast cancer cells, which was collected from GSE85871. One dataset can be loaded by the function `load_example_dataset()`, which is a list that includes expression data of 102 TCMs and the related metadata. The other one can be loaded by the function `data("data_logFC")`, which contains log2fc gene matrix performed by different expression analysis by limma. The third one can be loaded by the function `load_analyzedDEG()`, which contains the results of different expression analysis about each TCMs by using `deg_caller()`. The fourth one can be loaded by the function `load_AnalyzedSigPathway()`, which contains the significantly enriched pathways of each TCMs. The fifth one can be loaded by the function `data("tfdata")`, which contains the regulation realtionship of transcription factors and their target genes, which was collected from Msigdb.

# <a id="Section.4" style="color:#159957;">4. TCMR Pipeline</a>
## <a id="Section.4.1" style="color:#159957;">4.1 Pipeline Introduction</a>

* <div style="color:red">**DEG Module: get different expressed genes from different groups**</div> 
    * `deg_caller()`: used to run two group differential expression analysis.
    * `deg_batch_caller()`: used to run batch DEG analysis
</br>

* <div style="color:green">**Enrich Module: enrichment analysis from DEGs from TCM**</div> 
    * `get_sig_pathway()`: pathway enrichment analysis for differential expression analysis of TCM
    * `get_sig_GO()`: GO enrichment analysis for differential expression analysis of TCM
    * `getGSEA()`: run gene set enrichment analysis (GSEA) and generate a gsea figure.
    * `bar_plot()`: generate a barplot for special pathways or GO terms of TCM
    * `dotplot()`: generate a dotplot for special pathways or GO terms of TCM
    * `tf_enrich()`: transcription factors enrichment analysis to identify key transcription factors affected by TCM

</br>


* <div style="color:blue">**Signature Search Module: 5 signaturesearch methods(ss_xx series) to calculate similarity between TCM profiles and input geneset or profiles**</div> 
    * `ss_cmap()`: CMAP Search Method from Lamb et al
    * `ss_lincs()`: CMAP Search Method from Subramanian et al
    * `ss_gcmap()`: gCMAP.
    * `ss_cor()`: Correlation-based similarity metrics, such as Spearman or Pearson coefficients, can also be used as GESS methods
    * `ss_CoreGx()`: Correlation-based similarity metrics, such as Spearman or Pearson coefficients, can also be used as GESS methods
</br>

* <div style="color:orange">**Visualization Module: two plotting methods (circle heatmap and matrix dotplot) were used to visualize the special genesets in TCM profiles.**</div> 
    * `gene_plot()`: circle heatmap was used to visualize the special genesets in TCM profiles
    * `get_dot()`: matrix dotplot was used to visualize the special genesets in TCM profiles
</br>

## <a id="Section.4.2" style="color:#159957;">4.2 Steps of Pipeline</a>
Basically, the above three connected modules explain the workflow of this R package. MOVICS first identifies the cancer subtype (CS) by using one or multiple clustering algorithms; if multiple clustering algorithms are specified, it is highly recommended to perform a consensus clustering based on different subtyping results in order to derive stable and robust subtypes. Second, after having subtypes it is natural to exploit the heterogeneity of subtypes from as many angles as possible. Third, each subtype should have a list of subtype-specific biomarkers for reproducing such subtype in external cohorts. Therefore, let us follow this analytic pipeline and step into each module and each function in MOVICS. 

### <a id="Section.4.2.1" style="color:#159957;">4.2.1 DEG Module</a>
#### <a id="Section.4.2.1.1" style="color:#159957;">1) get data from example files</a>
Data used for this vignette should be first extracted from the function of `load_example_dataset()`, including expression data and pdata. **Notably, if you used your own data to perform different expression analysis, and please make sure your data structure is consistent with this.** 
```{r, eval=TRUE}
# print name of example tcm expression data
data <- load_example_dataset()
names(data)
data("tfdata")
```

### <a id="Section.4.2.1.2" style="color:#159957;">2) DEG analysis </a>

```{r, eval=TRUE}
# # Get a subset of expression dataset
ix <- c(1, 2, 61, 62)
expr <- data$expr[, ix]
group <- data$pdata$perturbagen[ix]
#' # Run DEG analysis
report <- deg_caller(expr, group = group, ref_group = group[3])
head(report)

# Run batch DEG analysis
reports <- deg_batch_caller(expr, groups = group, ref_group = group[3])
head(reports)
```

### <a id="Section.4.2.1.3" style="color:#159957;">3) enrichment analysis</a>

The most important parameter to estimate in any clustering study is the optimum number of clusters $k$ for the data, where $k$ needs to be small enough to reduce noise but large enough to retain important information. Herein MOVICS refers to CPI<sup>5</sup> and Gaps-statistics<sup>6</sup> to estimate the number of clusters by using `getClustNum()` function. 
```{r, fig.align="center", fig.width=6, fig.height=6, fig.cap="Figure 1. Identification of optimal cluster number by calculating CPI (blue line) and Gaps-statistics (red line) in TCGA-BRCA cohort.", eval=TRUE}
one_report <- TCMR::load_analyzedDEG(2)
msigdb.path <- system.file("extdata", "c2.cp.kegg.v6.2.symbols.gmt",package = "TCMR",mustWork = TRUE)
gsea.up <- get_GSEA(data = one_report,dirct = "up",msigdb.path = msigdb.path,n.path = 3,
         p.cutoff =0.05)
```
```{r, fig.align="center", fig.width=10, fig.height=8, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
gsea.up$fig
```

Check some columns of GSEA results for the first subtype (CS1).

```{r, eval=FALSE}
print(gsea.up$gsea.list[1:6,3:6])
```

```{r, echo=FALSE, eval=TRUE}
gsea.up$gsea.list[1:6,3:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```

#### <a id="Section.4.2.3.4" style="color:#159957;">4) run pathway enrichment analysis</a>
For all the new defined molecular subtypes, it is critical to depict its characteristics validated by different signatures of gene sets. MOVICS provides a simple function which uses gene set variation analysis to calculate enrichment score of each sample in each subtype based on given gene set list of interest. First, we must prepare a gene list of interest in hand saved as [GMT format](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29).

Then we can calculate single sample enrichment score based on specified method and given gene set of interest.
```{r,echo=FALSE, eval=TRUE}
# run GSVA to estimate single sample enrichment score based on given gene set of interest
data("AnalyzedDEG")
one_report <- load_analyzedDEG(1)
deg <- subset(one_report, abs(logFC) > 1 & P.Value < 0.05)
#path <- suppressMessages(get_sig_pathway(deg, p.cutoff = 1, p.adj.cutoff = 1, n.path = 10))

```

```{r, echo=TRUE, eval=TRUE}
path <- load_AnalyzedSigPathway(1)
```

```{r, echo=TRUE, eval=FALSE}
print(path)
```

```{r, echo=FALSE, eval=TRUE}
path[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```

```{r, fig.align="center", fig.width=12, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
bar_plot(path)
```

```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
dotplot(path)
```

```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
gene <- tfdata$gene[100:500]
tf_enrich(gene, tfdata = tfdata, colours = c("red", "blue"), colby = "p.adjust")
```

#### <a id="Section.4.2.1.3" style="color:#159957;">5) Signature search</a>

### <a id="Section.5" style="color:#159957;">1) Signature search</a>

#### <a id="Section.5" style="color:#159957;">1) CMAP</a>

```{r, eval=TRUE}
data("data_logFC")
upset <- rownames(data_logFC)[1:100]
downset <- rownames(data_logFC)[400:550]
input <- list(upset = upset, downset = downset)
```

```{r, eval=TRUE}
cmap_kk <- ss_cmap(input = input, data = data_logFC)
test <- head(cmap_kk, 20)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```


#### <a id="Section.5" style="color:#159957;">2) LINCS</a>

```{r, eval=TRUE}
data("data_logFC")
upset <- rownames(data_logFC)[1:100]
downset <- rownames(data_logFC)[400:550]
input <- list(upset = upset, downset = downset)
```

```{r, eval=TRUE}
lincs_kk <- ss_lincs(input, data_logFC)
test <- head(lincs_kk, 20)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```

#### <a id="Section.5" style="color:#159957;">3) gcmap</a>

```{r, eval=TRUE}
data("data_logFC")
query <- data_logFC[1:60, 1, drop = FALSE]
```

```{r, eval=TRUE}
gcmap_kk <- ss_gcmap(input = query, data = data_logFC)
test <- head(gcmap_kk, 20)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```


#### <a id="Section.5" style="color:#159957;">4) cor</a>

```{r, eval=TRUE}
data("data_logFC")
query <- data_logFC[1:60, 1, drop = FALSE]
```

```{r, eval=TRUE}
cor_kk <- ss_cor(query, data_logFC, method = "pearson")
test <- head(cor_kk, 20)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```


```{r, eval=TRUE}
cor_kk <- ss_cor(query, data_logFC, method = "spearman")
test <- head(cor_kk, 20)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```


```{r, eval=TRUE}
cor_kk <- ss_cor(query, data_logFC, method = "kendall")
test <- head(cor_kk, 20)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```



#### <a id="Section.5" style="color:#159957;">5) CoreGx</a>

```{r, eval=TRUE}
data("data_logFC")
query <- data_logFC[1:60, 1, drop = FALSE]
```

```{r, eval=TRUE}
CoreGx_kk <- ss_CoreGx(query, data_logFC[1:20])
test <- head(CoreGx_kk, 10)
```

```{r, eval=FALSE,echo=TRUE}
test[1:6,1:6]
```

```{r, eval=TRUE,echo=FALSE}
test[1:6,1:6] %>%
  kbl(caption = "Table 8. Demo of GSEA results for the first cancer subtype (CS1) of breast cancer in TCGA-BRCA cohort.") %>%
  kable_classic(full_width = TRUE, html_font = "Calibri")
```


```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
cs_plot(test, x = "raw_score", y = "tcm", colby = "pvalue", color = c("blue", "red"))
```



#### <a id="Section.6" style="color:#159957;">1) Visual </a>

```{r, eval=TRUE}
data("data_logFC")
```

```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
gene_plot(data = as.matrix(data_logFC[1:6, ]))
```

```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
gene_plot(data = as.matrix(data_logFC[1:6, ]), colors = c("blue", "white", "red"))
```

```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
gene_plot(data = as.matrix(data_logFC[1:10, ]), colors = c("skyblue", "white", "red"))
```

#### <a id="Section.6" style="color:#159957;">1) Dot </a>

```{r, eval=TRUE,echo=FALSE}
data <- as.matrix(data_logFC[1:6, ])
data <- data.frame(t(data), check.rows = FALSE, check.names = FALSE)
data$compound <- rownames(data)
data <- data[3:13, ]
```

```{r, fig.align="center", fig.width=7, fig.height=7, fig.cap="Figure 15. Heatmap of subtype-specific upregulated pathways using edgeR algorithm for 5 identified subtypes in TCGA-BRCA cohort.", eval=TRUE}
gene_dot(data)
```
